src/elfpy/markets.py

```diff
@@ -181,50 +115,57 @@ def trade_and_update(self, agent_action: MarketAction, pricing_model: PricingMod
         self.check_action_type(agent_action.action_type, pricing_model.model_name())
         # TODO: check the desired amount is feasible, otherwise return descriptive error
         # update market variables which may have changed since the user action was created
-        time_remaining = time_utils.get_yearfrac_remaining(self.time, agent_action.mint_time, self.token_duration)
-        stretched_time_remaining = time_utils.stretch_time(time_remaining, self.time_stretch_constant)
+        time_remaining = StretchedTime(
+            # TODO: We should improve the StretchedTime API so that it accepts yearfracs in
+            # the place of days remaining.
+            days=time_utils.get_yearfrac_remaining(
+                self.time, agent_action.mint_time, self.position_duration.normalized_days
+            )
+            * 365,
+            time_stretch=self.position_duration.time_stretch,
+        )
         # for each position, specify how to forumulate trade and then execute
         if agent_action.action_type == "open_long":  # buy to open long
             market_deltas, agent_deltas = self._open_long(
                 pricing_model=pricing_model,
                 agent_action=agent_action,
                 token_out="pt",
-                stretched_time_remaining=stretched_time_remaining,
+                time_remaining=time_remaining,
             )
         elif agent_action.action_type == "close_long":  # sell to close long
             market_deltas, agent_deltas = self._close_long(
                 pricing_model=pricing_model,
                 agent_action=agent_action,
                 token_out="base",
-                stretched_time_remaining=stretched_time_remaining,
+                time_remaining=time_remaining,
             )
         elif agent_action.action_type == "open_short":  # sell PT to open short
             market_deltas, agent_deltas = self._open_short(
                 pricing_model=pricing_model,
                 agent_action=agent_action,
                 token_out="pt",
-                stretched_time_remaining=stretched_time_remaining,
+                time_remaining=time_remaining,
             )
         elif agent_action.action_type == "close_short":  # buy PT to close short
             market_deltas, agent_deltas = self._close_short(
                 pricing_model=pricing_model,
                 agent_action=agent_action,
                 token_in="base",
-                stretched_time_remaining=stretched_time_remaining,
+                time_remaining=time_remaining,
             )
         elif agent_action.action_type == "add_liquidity":
             market_deltas, agent_deltas = self._add_liquidity(
                 pricing_model=pricing_model,
                 agent_action=agent_action,
-                time_remaining=time_remaining,
-                stretched_time_remaining=stretched_time_remaining,
+                time_remaining=time_remaining.normalized_days,
```

Comment:
I think this should receive the `time_remaining: StretchedTime` variable and then it can pull out `normalized days` & `stretched_time` within the function? The trade-off is that it is less explicit about how the argument is being used, but reducing the signature and keeping it consistent with the other actions is worth it imo.

Reply:
Yeah I agree, I was trying to avoid making these changes in this PR, but I'm probably going to go ahead and throw them in so that we can be done with them.